infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/AsyncStoragePodInterceptor.java

```diff
@@ -0,0 +1,154 @@
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.openshift.provision;
+
+import static java.lang.String.format;
+import static org.eclipse.che.workspace.infrastructure.kubernetes.namespace.pvc.EphemeralWorkspaceUtility.isEphemeral;
+import static org.eclipse.che.workspace.infrastructure.openshift.provision.AsyncStorageProvisioner.ASYNC_STORAGE;
+
+import io.fabric8.kubernetes.api.model.DoneablePod;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watch;
+import io.fabric8.kubernetes.client.Watcher;
+import io.fabric8.kubernetes.client.dsl.PodResource;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesInfrastructureException;
+import org.eclipse.che.workspace.infrastructure.openshift.OpenShiftClientFactory;
+import org.eclipse.che.workspace.infrastructure.openshift.environment.OpenShiftEnvironment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Checking if starting workspace configured to persisted storage and async storage pod is running.
+ * To prevent 'Multi-Attach error for volume' async storage pod will be shutdown. After deleting
+ * asynchronous storage pod procedure of starting workspace will be continued.
+ */
+public class AsyncStoragePodInterceptor {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AsyncStoragePodInterceptor.class);
+
+  private final OpenShiftClientFactory clientFactory;
+  private final String strategy;
+
+  @Inject
+  public AsyncStoragePodInterceptor(
+      @Named("che.infra.kubernetes.pvc.strategy") String strategy,
+      OpenShiftClientFactory openShiftClientFactory) {
+    this.strategy = strategy;
+    this.clientFactory = openShiftClientFactory;
+  }
+
+  public void intercept(OpenShiftEnvironment osEnv, RuntimeIdentity identity)
+      throws InfrastructureException {
+    if (!"common".equals(strategy)) {
+      return;
+    }
+
+    if (isEphemeral(osEnv.getAttributes())) {
+      return;
+    }
+
+    String namespace = identity.getInfrastructureNamespace();
+    PodResource<Pod, DoneablePod> podResource =
+        clientFactory
+            .create(identity.getWorkspaceId())
+            .pods()
+            .inNamespace(namespace)
+            .withName(ASYNC_STORAGE);
+    if (podResource.get() == null) {
+      return;
+    }
```

Comment:
@vparfonov I probably missing this part. It tries to find the Async storage pod and does nothing if it does not exist. Could you please clarify a bit. Maybe worth moving it to a separate method

Reply:
yep, make sense